src/common/walFilter/walFilter.c

```diff
@@ -0,0 +1,763 @@
+                main_state->tot_len = this->record->xl_tot_len;
+
+                result = true;
+                break;
+            }
+            else
+            {
+                bufUsedZero(buffer);
+                size = ioRead(storageReadIo(storageRead), buffer);
+                bufUsedSet(buffer, size);
+            }
+        }
+        lstClearFast(this->headers);
+    }
+    MEM_CONTEXT_TEMP_END();
+    return result;
+}
+
+static bool
+getEndOfRecord(WalFilterState *const this)
+{
+    MEM_CONTEXT_TEMP_BEGIN();
+
+    const String *walSegment = NULL;
+    const TimeLineID timeLine = this->current_header->xlp_tli;
+    uint64_t segno = this->current_header->xlp_pageaddr / this->segSize;
+    const String *const path = strNewFmt("%08X%08X", timeLine, (uint32) (segno / XLogSegmentsPerXLogId(this->segSize)));
+
+    const StringList *const segmentList = storageListP(
+        storageRepoIdx(this->archiveInfo->repoIdx),
+        strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(this->archiveInfo->archiveId), strZ(path)),
+        .expression = strNewFmt("^[0-f]{24}-[0-f]{40}" COMPRESS_TYPE_REGEXP "{0,1}$"));
+
+    if (strLstEmpty(segmentList))
+    {
+        // an exotic case where we couldn't even find the current file.
+        THROW(FormatError, "no WAL files were found in the repository");
+    }
+
+    uint64_t segnoDiff = UINT64_MAX;
+    for (uint32_t i = 0; i < strLstSize(segmentList); i++)
+    {
+        const String *file = strSubN(strLstGet(segmentList, i), 0, 24);
+        TimeLineID tli;
+        uint64_t fileSegNo = 0;
+        XLogFromFileName(strZ(file), &tli, &fileSegNo, this->segSize);
+
+        if (fileSegNo - segno < segnoDiff && fileSegNo > segno)
+        {
+            segnoDiff = fileSegNo - segno;
+            walSegment = strLstGet(segmentList, i);
+        }
+    }
+
+    // current file is newest
+    if (segnoDiff == UINT64_MAX)
+    {
+        memContextSwitchBack();
+        memContextDiscard();
+        return false;
+    }
+
+    const StorageRead *const storageRead = storageNewReadP(
+        storageRepoIdx(0),
+        strNewFmt(STORAGE_REPO_ARCHIVE "/%s/%s", strZ(this->archiveInfo->archiveId), strZ(walSegment)));
+
+    buildArchiveGetPipeLine(ioReadFilterGroup(storageReadIo(storageRead)), this->archiveInfo);
+
+    ioReadOpen(storageReadIo(storageRead));
+
+    Buffer *const buffer = bufNew(this->walPageSize);
+    size_t size = ioRead(storageReadIo(storageRead), buffer);
+    bufUsedSet(buffer, size);
+    while (this->got_len != this->record->xl_tot_len)
+    {
+        if (!readRecord(this, buffer))
+        {
+            if (ioReadEof(storageReadIo(storageRead)))
+            {
+                THROW(FormatError, "Unexpected WAL end");
+            }
+
+            bufUsedZero(buffer);
+            size = ioRead(storageReadIo(storageRead), buffer);
+            bufUsedSet(buffer, size);
+        }
+    }
+
+    MEM_CONTEXT_TEMP_END();
+    return true;
+}
```

Comment:
Large parts of the `readBeginOfRecord` and `getEndOfRecord` functions are identical. Why not separate them into one?